image/cas/interface.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cas implements generic content-addressable storage.
+package cas
+
+import (
+	"io"
+
+	"golang.org/x/net/context"
+)
+
+// Engine represents a content-addressable storage engine.
+//
+// This interface is for internal use of oci-image-tool for the time
+// being.  It is subject to change.  This notice will be removed when
+// and if the interface becomes stable.
+type Engine interface {
```

Comment:
On Mon, Aug 29, 2016 at 06:57:28PM -0700, Stephen Day wrote:
> > +// Engine represents a content-addressable storage engine.
> > +//
> > +// This interface is for internal use of oci-image-tool for the time
> > +// being.  It is subject to change.  This notice will be removed when
> > +// and if the interface becomes stable.
> > +type Engine interface {
> 
> If this interface is useful, I'd expect to see some generic
> algorithms in here. For example:
> 
> ```
> Copy(dstEngine, srcEngine Engine, digest string) error
> ```

That doesn't seem like an behavior that needs to be implemented by
each Engine implementation.  Instead, I'd rather add a helper that
implemented Copy on top of the generic Engine interface (and thus
worked with all implementations).  I already do this with
image/cas/put.go.  Should I add an image/cas/copy.go with your copy as
well?